torch/csrc/autograd/input_buffer.cpp

```diff
@@ -126,101 +78,195 @@ static void accumulate(
-      }
+  const auto device = var.device();
+  const auto device_type = device.type();
+  // TODO: Use at::accelerator::isAccelerator(device->type()) instead
+  bool is_accelerator =
+      device.is_cuda() || device.is_mtia() || device.is_privateuseone();
+  //
+  // [ Non-accelerator case ]
+  //
+  if (!is_accelerator) {
+    if (!buffer[pos].defined()) {
+      buffer[pos] = std::move(var);
     } else {
-      std::optional<c10::Stream> opt_sync_stream = std::nullopt;
-      const auto guard = c10::impl::VirtualGuardImpl{device_type};
-      if (on_consumer && !on_producer) {
-        // (3a)
-        opt_accumulate_stream = opt_consumer_stream;
-        opt_sync_stream = guard.getDefaultStream(opt_consumer_stream->device());
-      } else if (on_producer && !on_consumer) {
-        // (4a)
-        opt_accumulate_stream =
-            guard.getDefaultStream(opt_producer_stream->device());
-        opt_sync_stream = opt_producer_stream;
-      } else {
-        // (5)
-        opt_accumulate_stream = guard.getDefaultStream(*device);
-      }
-      if (opt_sync_stream && (opt_accumulate_stream != opt_sync_stream)) {
-        // (3b), (4b)
-        c10::OptionalDeviceGuard device_guard{opt_sync_stream->device()};
-        auto event = c10::Event{device_type};
-        event.record(*opt_sync_stream);
-        opt_accumulate_stream->wait(event);
-        const auto guard = c10::impl::VirtualGuardImpl(device_type);
-        record_stream_any_impl(var, *opt_accumulate_stream);
-      }
+      c10::OptionalDeviceGuard device_guard{device};
+      accumulate(buffer, pos, std::move(var));
     }
+    return;
   }
+  // Handle the case where var is on an accelerator but producer node has no
+  // canonical stream, e.g. this can happen if forward is DtoH
+  const std::optional<c10::Stream>& opt_producer_stream =
+      (opt_producer_stream_.has_value()
+           ? opt_producer_stream_
+           : std::optional<c10::Stream>(
+                 at::accelerator::getCurrentStream(device.index())));
 
-  auto& old_var = buffer[pos];
-  if (!old_var.defined()) {
+  TORCH_INTERNAL_ASSERT(opt_consumer_stream && opt_producer_stream);
+
+  // See Note: [Autograd Producer-Consumer Stream Syncs]
+  if (!opt_accum_streams[pos].has_value()) {
+    TORCH_INTERNAL_ASSERT(!buffer[pos].defined());
+    // [ First producer ]
+    // 1)
+    if (opt_consumer_stream->device() == device) {
+      // Case A
+      opt_accum_streams[pos] = opt_consumer_stream;
+    } else if (opt_producer_stream->device() == device) {
+      // Case B
+      opt_accum_streams[pos] = opt_producer_stream;
+    } else {
+      // Case C
+      opt_accum_streams[pos] =
+          at::accelerator::getCurrentStream(device.index());
+    }
+    // 2)
     buffer[pos] = std::move(var);
+    // 3)
+    ready_events[pos] = c10::Event{device_type};
+    ready_events[pos]->record(*opt_producer_stream);
+    ready_streams[pos] = opt_producer_stream;
```

Comment:
I would argue you don't need to record this "ready_streams" and just always sync on "redy_events" when it's there.

@ngimel should we be concerned about the perf impact of making a stream waiting on itself?

Reply:
This would be a good simplification indeed, although even if making a stream on waiting on itself were free, you may need a conditional on "ready_streams" in order to avoid extra record_stream. The issue is not that there would be any perf impact I think (it should be a no-op), but rather because we need to access a tensors data ptr in order to record stream and that is not always possible, e.g. for wrapper subclasses (unless there is a better way?).

Reply:
We'd rather avoid it for simple one-stream workloads which is still the majority

Reply:
Oh and we should definitely avoid extra record_streams

Reply:
Ok!